document/core/appendix/algorithm.rst

```diff
@@ -1,6 +1,193 @@
+
+.. note::
+   In the presentation of this algorithm, multiple values are supported for the :ref:`result types <syntax-resulttype>` classifying blocks and labels.
+   With the current version of WebAssembly, the :code:`list` could be simplified to an optional value.
+
+For the purpose of presenting the algorithm, the operand and control stacks are simply maintained as global variables:
+
+.. code-block:: none
+
+   var opds : opd_stack
+   var ctrls : ctrl_stack
+
+However, these variables are not manipulated directly by the main checking function, but through a set of auxiliary functions:
+
+.. code-block:: none
+
+   func push_opd(type : val_type | Unknown) =
+     opds.push(type)
+
+   func pop_opd() : val_type | Unknown =
+     if (opds.size() = ctrls[0].height && ctrls[0].unreachable) return Unknown
+     error_if(opds.size() = ctrls[0].height)
+     return opds.pop(type)
+
+   func pop_opd(expect : val_type | Unknown) : val_type | Unknown =
+     let actual = pop_opd()
+     if (actual = Unknown) return expect
+     if (expect = Unknown) return actual
+     error_if(actual =/= expect)
+     return actual
+
+   func push_opds(types : list(val_type)) = foreach (t in types) push_opd(t)
+   func pop_opds(types : list(val_type)) = foreach (t in reverse(types)) pop_opd(t)
+
+Pushing an operand simply pushes the respective type to the operand stack.
+
+Popping an operand checks that the operand stack does not underflow the current block and then removes one type.
+But first, a special case is handled where the block contains no known operands, but has been marked as unreachable.
+That can occur after an unconditional branch, when the stack is typed :ref:`polymorphically <polymorphism>`.
+In that case, an unknown type is returned.
+
+A second function for popping an operand takes an expected type, which the actual operand type is checked against.
+The types may differ in case one of them is Unknown.
+The more specific type is returned.
+
+Finally, there are accumulative functions for pushing or popping multiple operand types.
+
+The control stack is likewise manipulated through auxiliary functions:
+
+.. code-block:: none
+
+   func push_ctrl(label : list(val_type), out : list(val_type)) =
+     let frame = ctrl_frame(label, out, opds.size(), false)
+     ctrls.push(frame)
+
+   func pop_ctrl() : list(val_type) =
+     error_if(ctrls.is_empty())
+     let frame = ctrls.pop()
+     pop_opds(frame.end_types)
+     error_if(opds.size() =/= frame.height)
+     return frame.end_types
+
+   func unreachable() =
+     opds.resize(ctrl[0].height)
+     ctrl[0].unreachable := true
+
+Pushing a control frame takes the types of the label and result values.
```

Comment:
Might want to mention that `push` prepends. Otherwise the `br` steps below and `unreachable` above don't make sense.